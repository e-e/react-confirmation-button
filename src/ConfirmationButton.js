import React, { Component } from 'react';

class ConfirmButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      confirming: false,
      clicked: false
    };
    this.allClassNames = [
      'wrapClass',
      'buttonClass',
      'mainClass',
      'confirmClass',
      'cancelClass',
      'loadingClass',
      'disabledClass'
    ];

    this.initClasses();

    this.onConfirm = this.onConfirm.bind(this);
    this.onConfirmClicked = this.onConfirmClicked.bind(this);
    this.onCancel = this.onCancel.bind(this);
  }
  componentDidMount() {}
  initClasses() {
    this.allClassNames.forEach(name => {
      this[name] = '';
      if (
        typeof this.props[name] === 'string' &&
        this.props[name].trim().length
      ) {
        this[name] = `${this[name]} ${this.props[name]}`.trim();
      }
    });
  }
  outputClasses() {
    return [].slice.call(arguments).reduce((classStr, className) => {
      return `${classStr} ${className}`.trim();
    });
  }
  onConfirm() {
    const confirm =
      typeof this.props.onConfirm === 'function'
        ? this.props.onConfirm
        : () => {
            throw new Error(
              "[react-confirmation-button]: You must provide an 'onConfirm' function."
            );
          };
    confirm();
    this.setState({ confirming: false, clicked: true });
  }
  onCancel() {
    const cancel =
      typeof this.props.onCancel === 'function'
        ? this.props.onCancel
        : () => {};
    cancel();
    this.setState({ confirming: false });
  }
  onConfirmClicked() {
    this.setState({ confirming: true });
  }
  renderButton() {
    if (!!this.props.once && this.state.clicked) return null;
    if (this.state.confirming) return null;

    return (
      <button
        className={this.outputClasses(this.buttonClass, this.mainClass)}
        onClick={this.onConfirmClicked}
      >
        {this.props.buttonText || 'Save'}
      </button>
    );
  }
  renderConfirm() {
    if (!!this.props.once && this.state.clicked) return null;
    if (!this.state.confirming) return null;

    return [
      <button
        className={this.outputClasses(this.buttonClass, this.confirmClass)}
        onClick={this.onConfirm}
        key="confirm"
      >
        {this.props.confirmText || 'Confirm'}
      </button>,
      <button
        className={this.outputClasses(this.buttonClass, this.cancelClass)}
        onClick={this.onCancel}
        key="cancel"
      >
        {this.props.cancelText || 'Cancel'}
      </button>
    ];
  }

  renderDisabled() {
    if (!(this.state.clicked && !!this.props.once)) {
      return null;
    }
    return (
      <button
        className={this.outputClasses(this.buttonClass, this.disabledClass)}
        disabled
      >
        {this.props.loadingText || 'Loading'}
      </button>
    );
  }
  render() {
    return (
      <div className={this.wrapClass}>
        {this.renderButton()}
        {this.renderConfirm()}
        {this.renderDisabled()}
      </div>
    );
  }
}

export default ConfirmButton;
