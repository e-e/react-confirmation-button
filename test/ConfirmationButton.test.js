import React from 'react';
import ConfirmationButton from '../src/ConfirmationButton';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const component = renderer.create(
    <ConfirmationButton onConfirm={() => {}} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders correctly after clicking confirm', () => {
  const component = renderer.create(
    <ConfirmationButton onConfirm={() => {}} />
  );
  component.getInstance().onConfirmClicked();

  expect(component.toJSON()).toMatchSnapshot();
});

it('renders correctly with provided classnames', () => {
  const component = renderer.create(
    <ConfirmationButton
      onConfirm={() => {}}
      wrapClass="wrap-class"
      buttonClass="button-class"
      confirmClass="confirm-class"
      cancelClass="cancel-class"
    />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders correctly with alt button text', () => {
  const component = renderer.create(
    <ConfirmationButton
      onConfirm={() => {}}
      confirmText="Yes"
      buttonText="Do something"
      cancelText="No"
    />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders correctly after clicking confirm with alt button text', () => {
  const component = renderer.create(
    <ConfirmationButton
      onConfirm={() => {}}
      confirmText="Yes"
      buttonText="Do something"
      cancelText="No"
    />
  );
  component.getInstance().onConfirmClicked();

  expect(component.toJSON()).toMatchSnapshot();
});
