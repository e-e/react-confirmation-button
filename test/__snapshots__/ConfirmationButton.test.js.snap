// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className=""
>
  <button
    className=""
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;

exports[`renders correctly after clicking confirm 1`] = `
<div
  className=""
>
  <button
    className=""
    onClick={[Function]}
  >
    Confirm
  </button>
  <button
    className=""
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;

exports[`renders correctly after clicking confirm with alt button text 1`] = `
<div
  className=""
>
  <button
    className=""
    onClick={[Function]}
  >
    Yes
  </button>
  <button
    className=""
    onClick={[Function]}
  >
    No
  </button>
</div>
`;

exports[`renders correctly with alt button text 1`] = `
<div
  className=""
>
  <button
    className=""
    onClick={[Function]}
  >
    Do something
  </button>
</div>
`;

exports[`renders correctly with provided classnames 1`] = `
<div
  className="wrap-class"
>
  <button
    className="button-class"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
